File: numerical_analysis/example_code/topic2/auto.py
#  Autocorrelation time in the 2-D Ising Model

import math
import random

J = 1.0                     #  ferromagnetic coupling
L_x = L_y = 10              #  number of spins in x and y
N = L_x * L_y               #  number of spins
s = []                      #  the spins
T = 1.0                     #  temperature
H = 0.0                     #  magnetic field

w = []                      #  Boltzmann factors

def compute_Boltzmann_factors():
    global w
    w = []
    for m in range(5):
        w.append([])
        sum_of_neighbors = -4 + 2 * m
        for n in range(2):
            s_i = -1 + 2 * n
            factor = math.exp( -2.0 * (J * sum_of_neighbors + H) * s_i / T )
            w[m].append(factor)

eAv = mAv = 0.0             #  accumulators to compute <e> and <m>
nSave = 10                  #  values to save for autocorrelations
eSave = []; mSave = []      #  saved energy and magnetization values
cee = []; cmm = []          #  energy and magnetization correlation sums
nCorr = 0                   #  number of values accumulated in sums

def initialize_correlations():
    global eAv, mAv, eSave, mSave, cee, cmm, nCorr
    eAv = mAv = 0.0
    eSave.clear()
    mSave.clear()
    cee = [0.0] * (nSave + 1)
    cmm = [0.0] * (nSave + 1)
    nCorr = 0

steps = 0                  #  steps so far

def initialize():
    # create spin lattice and set spin randomly up or down (hot start)
    global s, steps
    s = []
    for i in range(L_x):
        s.append([])
        for j in range(L_y):
            s[i].append(random.choice((-1, 1)))
    compute_Boltzmann_factors()
    steps = 0

def Metropolis_step_accepted():

    # choose a random spin
    i = random.randrange(L_x)
    j = random.randrange(L_y)

    # find the sum of neighbors assuming periodic boundary conditions
    sum_of_neighbors = ( s[(i-1)%L_x][j] + s[(i+1)%L_x][j] +
                         s[i][(j-1)%L_y] + s[i][(j+1)%L_y] )

    # access ratio of precomputed Boltzmann factors
    ratio = w[2 + int(sum_of_neighbors/2)][int((1 + s[i][j])/2)]

    # apply the Metropolis test
    if ratio > 1.0 or ratio > random.random():
        s[i][j] = -s[i][j]
        return True
    else:
        return False

acceptance_ratio = 0        # accepted steps / total number of steps

def one_Monte_Carlo_step_per_spin():
    global acceptance_ratio, steps
    accepts = 0
    for n in range(N):
        if Metropolis_step_accepted():
            accepts += 1
    acceptance_ratio = accepts / float(N)
    steps += 1

def magnetization_per_spin():
    s_sum = 0.0
    for i in range(L_x):
        for j in range(L_y):
            s_sum += s[i][j]
    return s_sum / float(N)

def energy_per_spin():
    s_sum = 0.0
    ss_sum = 0.0
    for i in range(L_x):
        for j in range(L_y):
            s_sum += s[i][j]
            ss_sum += s[i][j] * (s[(i+1)%L_x][j] + s[i][(j+1)%L_y])
    return -(J * ss_sum + H * s_sum) / float(N)

def accumulate_correlations():
    global nCorr, eAv, mAv, cee, cmm, eSave, mSave
    #  calculate current energy and magnetization
    e = energy_per_spin()
    m = magnetization_per_spin()
    #  accumulate averages and correlation products
    if len(eSave) == nSave:     #  if nSave values have been saved
        nCorr += 1
        eAv += e
        mAv += m
        cee[0] += e**2
        cmm[0] += m**2
        for i in range(1, nSave + 1):
            cee[i] += eSave[i - 1] * e
            cmm[i] += mSave[i - 1] * m
        #  discard the oldest values
        eSave.pop()
        mSave.pop()
    #  save the current values
    eSave.insert(0, e)
    mSave.insert(0, m)

tau_e = tau_m = 0.0         #  autocorrelation times

def compute_autocorrelation_times():
    global tau_e, tau_m
    #  energy correlation
    av = eAv / nCorr
    c0 = cee[0] / nCorr - av**2
    tau_e = 0.0
    for i in range(nSave + 1):
        tau_e += (cee[i] / nCorr - av**2) / c0
    #  magnetization correlation
    av = mAv / nCorr
    c0 = cmm[0] / nCorr - av**2
    tau_m = 0.0
    for i in range(nSave + 1):
        tau_m += (cmm[i] / nCorr - av**2) / c0

print(" Two-dimensional Ising Model - Autocorrelation times")
print(" ---------------------------------------------------")
L_x = L_y = int(input(" Enter number of spins L in each direction: "))
N = L_x * L_y
T1 = float(input(" Enter starting temperature: "))
T2 = float(input(" Enter ending temperature: "))
TSteps = int(input(" Enter number of temperature steps: "))
MCSteps = int(input(" Enter number of Monte Carlo steps: "))

initialize()
file = open("auto.data", "w")
thermSteps = int(0.2 * MCSteps)
for i in range(TSteps + 1):
    T = T1 + i * (T2 - T1) / float(TSteps)
    compute_Boltzmann_factors()
    for step in range(thermSteps):
        one_Monte_Carlo_step_per_spin()
    initialize_correlations()
    for step in range(MCSteps):
        one_Monte_Carlo_step_per_spin()
        accumulate_correlations()
    compute_autocorrelation_times()
    print(" T = ", str(T), "\ttau_e = ", str(tau_e), "\ttau_m = ", str(tau_m))
    file.write(str(T) + '\t' + str(tau_e) + '\t' + str(tau_m) + '\n')
file.close()

File: example_code/topic2/auto.cpp
// Autocorrelation time in the 2-D Ising Model

#include <cmath>
#include <cstdlib>
#include <iostream>
#include <fstream>
#include <list>                 // to save values for autocorrelations
using namespace std;

inline double std_rand()
{
    return rand() / (RAND_MAX + 1.0);
}

double J = +1;                  // ferromagnetic coupling
int Lx, Ly;                     // number of spins in x and y
int N;                          // number of spins
int **s;                        // the spins
double T;                       // temperature
double H = 0;                   // magnetic field

double w[17][3];                // Boltzmann factors

void computeBoltzmannFactors ( ) {
     for (int i = -8; i <= 8; i += 4) {
          w[i + 8][0] = exp( - (i * J + 2 * H) / T);
          w[i + 8][2] = exp( - (i * J - 2 * H) / T);
     }
}

double eAv, mAv;                // accumulators to compute <e> and <m>
int nSave = 10;                 // values to save for autocorrelations
list<double> eSave, mSave;      // saved energy and magnetization values
double *cee, *cmm;              // energy and magnetization correlation sums
int nCorr;                      // number of values accumulated in sums

void initializeCorrelations() {
     eAv = mAv = 0;
     eSave.clear();
     mSave.clear();
     if (cee != NULL) delete [] cee;
     if (cmm != NULL) delete [] cmm;
     cee = new double [nSave + 1];
     cmm = new double [nSave + 1];
     for (int i = 0; i <= nSave; i++)
          cee[i] = cmm[i] = 0;
     nCorr = 0;
}

int steps = 0;                  // steps so far

void initialize ( ) {
     s = new int* [Lx];
     for (int i = 0; i < Lx; i++)
          s[i] = new int [Ly];
     for (int i = 0; i < Lx; i++)
          for (int j = 0; j < Ly; j++)
               s[i][j] = std_rand() < 0.5 ? +1 : -1;   // hot start
     computeBoltzmannFactors();
     steps = 0;
}

bool MetropolisStep ( ) {

     // choose a random spin
     int i = int(Lx*std_rand());
     int j = int(Ly*std_rand());

     // find its neighbors using periodic boundary conditions
     int iPrev = i == 0 ? Lx-1 : i-1;
     int iNext = i == Lx-1 ? 0 : i+1;
     int jPrev = j == 0 ? Ly-1 : j-1;
     int jNext = j == Ly-1 ? 0 : j+1;

     // find sum of neighbors
     int sumNeighbors = s[iPrev][j] + s[iNext][j] + s[i][jPrev] + s[i][jNext];
     int delta_ss = 2*s[i][j]*sumNeighbors;

     // ratio of Boltzmann factors
     double ratio = w[delta_ss+8][1+s[i][j]];
     if (std_rand() < ratio) {
          s[i][j] = -s[i][j];
          return true;
     } else return false;
}

double acceptanceRatio;

void oneMonteCarloStepPerSpin ( ) {
     int accepts = 0;
     for (int i = 0; i < N; i++)
          if (MetropolisStep())
               ++accepts;
     acceptanceRatio = accepts/double(N);
     ++steps;
}

double magnetizationPerSpin ( ) {
     int sSum = 0;
     for (int i = 0; i < Lx; i++)
     for (int j = 0; j < Ly; j++) {
          sSum += s[i][j];
     }
     return sSum / double(N);
}

double energyPerSpin ( ) {
     int sSum = 0, ssSum = 0;
     for (int i = 0; i < Lx; i++)
     for (int j = 0; j < Ly; j++) {
          sSum += s[i][j];
          int iNext = i == Lx-1 ? 0 : i+1;
          int jNext = j == Ly-1 ? 0 : j+1;
          ssSum += s[i][j]*(s[iNext][j] + s[i][jNext]);
     }
     return -(J*ssSum + H*sSum)/N;
}

void accumulateCorrelations() {

     // calculate current energy and magnetization
     double e = energyPerSpin();
     double m = magnetizationPerSpin();

     // accumulate averages and correlation products
     if (eSave.size() == nSave) {   // if nSave values have been saved
          ++nCorr;
          eAv += e;
          mAv += m;
          cee[0] += e * e;
          cmm[0] += m * m;
          list<double>::const_iterator ie = eSave.begin(), im = mSave.begin();
          for (int i = 1; i <= nSave; i++) {
               cee[i] += *ie++ * e;
               cmm[i] += *im++ * m;
          }

          // discard the oldest values
          eSave.pop_back();
          mSave.pop_back();
     }

     // save the current values
     eSave.push_front(e);
     mSave.push_front(m);
}

double tau_e, tau_m;

void computeAutocorrelationTimes() {

     // energy correlation
     double av = eAv / nCorr;
     double c0 = cee[0] / nCorr - av * av;
     tau_e = 0;
     for (int i = 1; i <= nSave; i++)
          tau_e += (cee[i] / nCorr - av * av) / c0;

     // magnetization correlation
     av = mAv / nCorr;
     c0 = cmm[0] / nCorr - av * av;
     tau_m = 0;
     for (int i = 1; i <= nSave; i++)
          tau_m += (cmm[i] / nCorr - av * av) / c0;
}

int main (int argc, char *argv[]) {

     cout << " Two-dimensional Ising Model - Autocorrelation times\n"
          << " ---------------------------------------------------\n"
          << " Enter number of spins L in each direction: ";
     cin >> Lx;
     Ly = Lx;
     N = Lx * Ly;
     double T1, T2;
     cout << " Enter starting temperature: ";
     cin >> T1;
     cout << " Enter ending temperature: ";
     cin >> T2;
     cout << " Enter number of temperature steps: ";
     int TSteps;
     cin >> TSteps;
     cout << " Enter number of Monte Carlo steps: ";
     int MCSteps;
     cin >> MCSteps;

     initialize();
     ofstream file("auto.data");
     int thermSteps = int(0.2 * MCSteps);
     for (int i = 0; i <= TSteps; i++) {
          T = T1 + i * (T2 - T1) / double(TSteps);
          computeBoltzmannFactors();
          for (int s = 0; s < thermSteps; s++)
               oneMonteCarloStepPerSpin();
          initializeCorrelations();
          for (int s = 0; s < MCSteps; s++) {
               oneMonteCarloStepPerSpin();
               accumulateCorrelations();
          }
          computeAutocorrelationTimes();
          cout << " T = " << T << "\ttau_e = " << tau_e
               << "\ttau_m = " << tau_m << endl;
          file << T << '\t' << tau_e << '\t' << tau_m << '\n';
     }
     file.close();
}
